common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+ * called in a loop until it returns false.
+ *
+ * Currently, we only send CMDU messages through message-oriented sockets and there are no plans
+ * to use message-stream sockets for it so, for the moment, only one implementation is required
+ * (the easy one).
+ *
+ * This interface and its implementations allow the separation of the logic around sockets and the
+ * logic around message parsing. The goal of this separation is to be able to test message parsing
+ * without having a peer connected at the other end of the wire, sending messages. During testing,
+ * we can fill the buffer with any contents we want and pretend that those bytes have been
+ * received through the socket connection. This way it is possible to test the parser with very odd
+ * combinations of data that would otherwise be very difficult to set and reproduce. The only limit
+ * for the tests we can write is our imagination.
+ */
+class CmduParser {
+public:
+    virtual ~CmduParser() = default;
+    /**
+     * Parses a CMDU out of a byte buffer.
+     *
+     * @param buffer array of bytes containing the message to parse.
+     * @param length number of bytes in the buffer.
+     * @param[out] cmdu_rx parsed message
+     * @return True if a message could be parsed out of the buffer and false otherwise (i.e.:
+     * because only a fragment of the message was given)
+     */
+    virtual bool parse_message(const sockets::Buffer &buffer, size_t length,
+                               ieee1905_1::CmduMessageRx &cmdu_rx) = 0;
+};
+
+class CmduParserImpl : public CmduParser {
+public:
+    bool parse_message(const sockets::Buffer &buffer, size_t length,
+                       ieee1905_1::CmduMessageRx &cmdu_rx) override
+    {
+        // TODO
+        return true;
+    }
+};
+
+/**
+ * This class is used to receive, parse and handle CMDU messages.
+ *
+ * This class is **testable** since all of its dependencies are injected in the constructor and is
+ * coded against interfaces (it is not tied to concrete implementations).
+ *
+ * An identical class should be created to process other kind of messages like, for example, UCC
+ * string configuration commands. Consider creating a template class instead of copying and
+ * pasting code.
+ */
+class CmduProcessor {
+public:
+    /**
+     * Class constructor.
+     *
+     * @param connection Socket connection on which CMDU messages are received.
+     * @param loop Event loop (reactor design pattern) to wait for read events on socket connection.
+     * @param parser CMDU message parser to build messages out of a sequence of bytes.
+     * @param handler CMDU message handler to handle received messages.
+     */
+    CmduProcessor(const std::shared_ptr<sockets::Connection> &connection,
+                  const std::shared_ptr<EventLoop<sockets::FileDescriptor>> &loop,
+                  const std::shared_ptr<CmduParser> &parser,
+                  const std::shared_ptr<CmduHandler> &handler)

```

Comment:
I can't think of any situation where you would need to keep any of these parameters alive longer than the CmduProcessor itself. So I'd pass them as (non-const) reference and let the CmduProcessor delete them when done. But maybe shared_ptr is a simpler way to express this.

I still feel that C++ is missing a convenient way to create an object in one function and pass its ownership to another.

Reply:
isn't `std::move` exactly for that purpose?

Reply:
std::move doesn't work with polymorphic objects. I.e. you can't do:
```
class CmduProcessor {
    CmduProcessor(socket::Connection &&connection);
}

MyConcreteConnection c;
CmduProcessor p(std::move(c));
```
because the abstract class Connection is not "move"able itself. And if it would have a move constructor, the result would be a Connection, not a MyConcreteConnection.

Also, we often want to still be able to use the object in the caller afterwards.